ext/oboe_metal/src/profiling.cc

```diff
@@ -0,0 +1,405 @@
+
+    // check if this thread is being profiled
+    if (prof_data_map[tid].running_p) {
+        // executes in the same thread as rb_postponed_job was called from
+
+        // get the frames
+        // won't overrun frames buffer, because size is set in arg 2
+        int num = rb_profile_frames(0, sizeof(frames_buffer) / sizeof(VALUE), frames_buffer, lines_buffer);
+
+        Profiling::process_snapshot(frames_buffer, num, tid, ts);
+    }
+
+    // add this timestamp as omitted to other running threads that are profiled
+    for (pair<const pid_t, prof_data_t> &ele : prof_data_map) {
+        if (ele.second.running_p && ele.first != tid) {
+            frames_buffer[0] = PR_OTHER_THREAD;
+            Profiling::process_snapshot(frames_buffer, 1, ele.first, ts);
+        }
+    }
+}
+
+void Profiling::profiler_record_gc() {
+   pid_t tid = AO_GETTID;
+   long ts = ts_now();
+
+    // check if this thread is being profiled
+    if (prof_data_map[tid].running_p) {
+        frames_buffer[0] = PR_IN_GC;
+        Profiling::process_snapshot(frames_buffer, 1, tid, ts);
+    }
+
+    // add this timestamp as omitted to other running threads that are profiled
+    for (pair<const pid_t, prof_data_t> &ele : prof_data_map) {
+        if (ele.second.running_p && ele.first != tid) {
+            frames_buffer[0] = PR_OTHER_THREAD;
+            Profiling::process_snapshot(frames_buffer, 1, ele.first, ts);
+        }
+    }
+}
+
+void Profiling::send_omitted(pid_t tid, long ts) {
+    static vector<FrameData> empty;
+    Logging::log_profile_snapshot(prof_data_map[tid].md,
+                                  prof_data_map[tid].prof_op_id,
+                                  ts,                              // timestamp
+                                  empty,                           // <vector> new frames
+                                  0,                               // number of exited frames
+                                  prof_data_map[tid].prev_num,     // total number of frames
+                                  prof_data_map[tid].omitted,      // array of timestamps of omitted snapshots
+                                  prof_data_map[tid].omitted_num,  // number of omitted snapshots
+                                  tid);                            // thread id
+
+    prof_data_map[tid].omitted_num = 0;
+}
+
+void Profiling::process_snapshot(VALUE *frames_buffer, int num, pid_t tid, long ts) {
+    int num_new = 0;
+    int num_exited = 0;
+    vector<FrameData> new_frames;
+
+    num = Frames::remove_garbage(frames_buffer, num);
+
+    // find the number of matching frames from the top
+    int num_match = Frames::num_matching(frames_buffer,
+                                         num,
+                                         prof_data_map[tid].prev_frames_buffer,
+                                         prof_data_map[tid].prev_num);
+    num_new = num - num_match;
+
+    num_exited = prof_data_map[tid].prev_num - num_match;
```

Comment:
this private function only gets called if `prof_data_map[tid]` exists

Reply:
This might be the case right now. But it's generally a good practice to not rely on outside checks. For example future code modifications could change this behaviour.